File: LeetCode/0653. Two Sum IV - Input is a BST/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

"""
422 / 422 test cases passed.
Runtime: 28 ms
Memory Usage: 37.9 MB
"""
class Solution:
    def findTarget(self, root: Optional[TreeNode], k: int) -> bool:
        if not root: return False
        que = collections.deque([root]);
        rec = set()
        while que:
            node = que.popleft()
            if (k - node.val) in rec:
                return True
            rec.add(node.val)
            if node.left: que.append(node.left)
            if node.right: que.append(node.right)
        return False

File: LeetCode/0653. Two Sum IV - Input is a BST/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

/**
 * 422 / 422 test cases passed.
 * Runtime: 28 ms
 * Memory Usage: 37.9 MB 
 */
class Solution {
public:
    unordered_set<int> rec;
    bool findTarget(TreeNode* root, int k) {
        if (!root) return false;
        if (rec.count(k - root->val)) return true;
        rec.insert(root->val);
        return findTarget(root->left, k) || findTarget(root->right, k);
    }
};
